src/survey/ia-feedback-survey.ts

```diff
@@ -0,0 +1,843 @@
+      >
+        <div
+          class="focus-trap"
+          tabindex="0"
+          @focus=${this.focusSubmitButton}
+        ></div>
+        <div
+          id="popup"
+          role="dialog"
+          aria-modal="true"
+          aria-labelledby="survey-heading"
+          style="left: ${this.popupTopX}px; top: ${this.popupTopY}px"
+        >
+          <h2 id="survey-heading" class="sr-only">${msg('Feedback Survey')}</h2>
+          <form
+            id="form"
+            ?disabled=${shouldDisableControls}
+            @click=${(e: Event) => e.stopPropagation()}
+            @keydown=${(e: Event) => e.stopPropagation()}
+            @submit=${this.submit}
+          >
+            <slot id="questions-slot"></slot>
+            ${this.error ? html`<div id="error">${this.error}</div>` : nothing}
+            <div id="actions">
+              <button
+                type="button"
+                id="cancel-button"
+                class="cta-button"
+                tabindex="0"
+                ?disabled=${shouldDisableControls}
+                @click=${this.cancel}
+              >
+                ${msg('Cancel')}
+              </button>
+              <input
+                type="submit"
+                id="submit-button"
+                class="cta-button"
+                tabindex="0"
+                ?disabled=${shouldDisableControls}
+                .value=${submitButtonText}
+              />
+            </div>
+          </form>
+        </div>
+        <div
+          class="focus-trap"
+          tabindex="0"
+          @focus=${this.focusFirstFormElement}
+        ></div>
+      </div>
+    `;
+  }
+
+  /**
+   * Assigns focus to the first slotted element in the form.
+   */
+  private focusFirstFormElement(): void {
+    this.assignedElements[0]?.focus();
+  }
+
+  /**
+   * Assigns focus to the submit button at the end of the form.
+   */
+  private focusSubmitButton(): void {
+    this.submitButton?.focus();
+  }
+
+  /**
+   * Handler for when the background layer outside the popup is clicked, treated
+   * as though the Cancel button were clicked.
+   */
+  private backgroundClicked(e: MouseEvent) {
+    if (!(e.target instanceof Node)) return;
+    if (this.popup?.contains(e.target)) return;
+    this.cancel(e);
+  }
+
+  /**
+   * Closes the survey popup and resets any submission error state.
+   */
+  private cancel(e: Event) {
+    e.preventDefault();
+    this.closePopup();
+    if (this.submissionState !== 'submitted') this.resetSubmissionState();
+  }
+
+  /**
+   * Returns a boolean indicating whether all required survey questions have received a
+   * valid response.
+   *
+   * - Questions with `required: true` and `type: 'vote'` must have an up/down vote recorded.
+   * - Questions with `required: true` and `type: 'comment'` must have a comment recorded.
+   * - Questions without `required: true`, or having `type: 'extra'`, are always valid.
+   */
+  private validate(): boolean {
```

Comment:
Very nice!

Reply:
This is a good reminder that I need to update that doc comment, since questions are now standalone components rather than objects specifying a `type` (a relic of a previous iteration) 😅